File: Python/15. 3Sum.py
from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        lists = []
        nums.sort()
        
        for i, a in enumerate(nums):
            if i>0 and a==nums[i-1]:
                continue
            
            j,k = i+1, len(nums)-1
            
            while j<k:
                tsum = a + nums[j] + nums[k]
                if tsum>0:
                    k-=1
                elif tsum<0:
                    j+=1
                else:
                    lists.append([a,nums[j],nums[k]])
                    j+=1
                    while nums[j]==nums[j-1] and j<k:
                        j+=1
        
        return lists
    
if __name__ == '__main__':
    nums = [-1,0,1,2,-1,-4]
    result = Solution().threeSum(nums)
    print(result)

File: CPP/15. 3Sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include "printFunctions.hpp"

using namespace std;

class Solution {
public:
	vector<vector<int>> threeSum(vector<int>& nums) {
		vector<vector<int>> triplets;

		sort(nums.begin(), nums.end());

		for (int i = 0; i < nums.size(); i++) {
			if (i > 0 and nums[i] == nums[i - 1]) {
				continue;
			}
			int j = i + 1, k = nums.size() - 1;

			while (j < k) {
				int sum = nums[i] + nums[j] + nums[k];

				if (sum > 0) {
					k--;
				}
				else if (sum < 0) {
					j++;
				}
				else {
					vector<int> ans = { nums[i], nums[j], nums[k] };
					triplets.push_back(ans);
					j++;
				}
			}
		}

		return triplets;
	}
};

int main() {
	Solution sol;
	vector<int> nums = { -1, 0, -1, 2, -1, -4 };
	vector<vector<int>> result = sol.threeSum(nums);
	printVectorOfVectors(result);
	return 0;
}